dbus/dbus_libgio_test.go

```diff
@@ -35,3 +48,295 @@ func TestIsGUID(t *testing.T) {
+				return false, nil
+			},
+			outBoolean: false,
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			gid, err = libgio.BusRegisterInterface(conn, tc.path, tc.xml)
+			assert.NoError(t, err)
+			assert.Greater(t, gid, uint(0))
+
+			libgio.RegisterMethodCallCallback(tc.path, tc.interfaceName, tc.methodName, tc.callback)
+
+			loop := libgio.MainLoopNew()
+			go libgio.MainLoopRun(loop)
+			defer libgio.MainLoopQuit(loop)
+
+			// let the dbus-daemon to set up
+			time.Sleep(500 * time.Millisecond)
+
+			// client code, call the dbus method, isolated from the code above
+			func() {
+				if tc.outString != "" {
+					var value string
+					interfaceMethodName := fmt.Sprintf("%s.%s", tc.interfaceName, tc.methodName)
+					err = godbusConn.Object(objectName, godbus.ObjectPath(tc.path)).Call(interfaceMethodName, 0).Store(&value)
+					assert.NoError(t, err)
+					assert.Equal(t, tc.outString, value)
+				} else {
+					var value bool
+					interfaceMethodName := fmt.Sprintf("%s.%s", tc.interfaceName, tc.methodName)
+					err = godbusConn.Object(objectName, godbus.ObjectPath(tc.path)).Call(interfaceMethodName, 0).Store(&value)
+					assert.NoError(t, err)
+					assert.Equal(t, tc.outBoolean, value)
+				}
+			}()
+		})
+
+		// let the dbus-daemon to clean up
+		time.Sleep(500 * time.Millisecond)
+	}
+}
+
+func TestMainLoop(t *testing.T) {
+	loop := libgio.MainLoopNew()
+	go libgio.MainLoopRun(loop)
+	defer libgio.MainLoopQuit(loop)
+
+	time.Sleep(100 * time.Millisecond)
+}
+
+func TestEmitSignal(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+
+	conn, err := libgio.BusGet(GBusTypeSystem)
+	assert.NoError(t, err)
+	assert.NotNil(t, conn)
+
+	const objectName = "io.mender.AuthenticationManager"
+
+	gid, err := libgio.BusOwnNameOnConnection(conn, objectName,
+		DBusNameOwnerFlagsAllowReplacement|DBusNameOwnerFlagsReplace)
+	assert.NoError(t, err)
+	assert.Greater(t, gid, uint(0))
+
+	godbusConn, err := godbus.SystemBus()
+	assert.NoError(t, err)
+	defer godbusConn.Close()
+
+	xml := `<node>
+	<interface name="io.mender.AuthenticationManager1">
+		<signal name="ValidJwtTokenAvailable">
+			<arg type="b" name="success" direction="out"/>
+		</signal>
+	</interface>
+</node>`
+
+	testCases := map[string]struct {
+		objectPath    string
+		interfaceName string
+		signalName    string
+		err           error
+	}{
+		"ok": {
+			objectPath:    "/io/mender/AuthenticationManager/TestEmitSignal1",
+			interfaceName: "io.mender.AuthenticationManager1",
+			signalName:    "ValidJwtTokenAvailable",
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			gid, err = libgio.BusRegisterInterface(conn, tc.objectPath, xml)
+			assert.NoError(t, err)
+			assert.Greater(t, gid, uint(0))
+
+			loop := libgio.MainLoopNew()
+			go libgio.MainLoopRun(loop)
+			defer libgio.MainLoopQuit(loop)
+
+			// let the dbus-daemon to set up
+			time.Sleep(500 * time.Millisecond)
```

Comment:
I'm no fan of sleeps. This is what typically causes strange failures six months down the road. Isn't there a better way to check this? Maybe using [`assert.Eventually`](https://godoc.org/github.com/stretchr/testify/assert#Eventually) with a longer timeout? Or perhaps you could get the progress of setup and teardown through channels?

Reply:
This is not needed anymore, it was part of my "debugging" because I thought we have issues here but the test failed in any case. The tests bass without any sleep, thus I'm removing them.